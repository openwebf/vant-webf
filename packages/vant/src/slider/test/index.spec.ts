import { Slider } from '..';
import {
  later,
  mount,
  trigger,
  triggerDrag,
  mockGetBoundingClientRect,
} from '../../../test';

function mockRect(vertical?: boolean) {
  return mockGetBoundingClientRect({
    width: vertical ? 0 : 100,
    height: vertical ? 100 : 0,
    top: vertical ? 0 : 100,
    left: vertical ? 100 : 0,
  });
}

test('should emit "update:modelValue" event after dragging button', () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
    },
  });

  const button = wrapper.find('.van-slider__button');
  triggerDrag(button, 50, 0);
  expect(wrapper.emitted('update:modelValue')!.pop()).toEqual([100]);
});

test('should emit "update:modelValue" event after clicking slider', () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
    },
  });

  trigger(wrapper, 'click', 100, 0);
  expect(wrapper.emitted('update:modelValue')!.pop()).toEqual([100]);
});

test('should emit drag-start event when start dragging', () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
    },
  });

  const button = wrapper.find('.van-slider__button');
  trigger(button, 'touchstart');
  trigger(button, 'touchmove');
  expect(wrapper.emitted('drag-start')).toBeTruthy();
});

test('should emit drag-end event when end dragging', () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
    },
  });

  const button = wrapper.find('.van-slider__button');
  trigger(button, 'touchstart');
  trigger(button, 'touchmove');
  expect(wrapper.emitted('drag-end')).toBeFalsy();
  trigger(button, 'touchend');
  expect(wrapper.emitted('drag-end')).toBeTruthy();
});

test('should not allow to drag slider when disabled', async () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
      disabled: true,
    },
  });

  const button = wrapper.find('.van-slider__button');
  triggerDrag(button, 50, 0);
  expect(wrapper.emitted('update:modelValue')).toBeFalsy();
});

test('should not allow to click slider when disabled', async () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
      disabled: true,
    },
  });

  trigger(wrapper, 'click', 100, 0);
  expect(wrapper.emitted('update:modelValue')).toBeFalsy();
});

test('should not allow to drag slider when readonly', async () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
      readonly: true,
    },
  });

  const button = wrapper.find('.van-slider__button');
  triggerDrag(button, 50, 0);
  expect(wrapper.emitted('update:modelValue')).toBeFalsy();
});

test('should not allow to click slider when readonly', async () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
      readonly: true,
    },
  });

  trigger(wrapper, 'click', 100, 0);
  expect(wrapper.emitted('update:modelValue')).toBeFalsy();
});

test('should render readonly Slider correctly', async () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
      readonly: true,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should allow to drag vertical slider', () => {
  const restoreMock = mockRect(true);

  const wrapper = mount(Slider, {
    props: {
      vertical: true,
      modelValue: 50,
    },
  });

  const button = wrapper.find('.van-slider__button');
  triggerDrag(button, 0, 50);
  expect(wrapper.emitted('update:modelValue')!.pop()).toEqual([100]);

  restoreMock();
});

test('should change slider bar height when using bar-height prop', () => {
  const wrapper = mount(Slider, {
    props: {
      barHeight: 10,
      modelValue: 50,
    },
  });

  expect(wrapper.style.height).toEqual('10px');
});

test('should change button size when using button-size prop', () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
      buttonSize: 10,
    },
  });

  const button = wrapper.find('.van-slider__button');
  expect(button.style.width).toEqual('10px');
  expect(button.style.height).toEqual('10px');
});

test('should emit "update:modelValue" event after clicking vertical slider', () => {
  const wrapper = mount(Slider, {
    props: {
      vertical: true,
      modelValue: 50,
    },
  });

  trigger(wrapper, 'click', 0, 100);
  expect(wrapper.emitted('update:modelValue')!.pop()).toEqual([100]);
});

test('should not emit change event when value not changed', async () => {
  const wrapper = mount(Slider, {
    props: {
      modelValue: 50,
    },
  });

  const button = wrapper.find('.van-slider__button');
  trigger(button, 'touchstart');
  trigger(wrapper, 'click', 100, 0);
  expect(wrapper.emitted('change')).toHaveLength(1);

  await wrapper.setProps({ modelValue: 100 });
  trigger(button, 'touchstart');
  trigger(wrapper, 'click', 100, 0);

  expect(wrapper.emitted('change')).toHaveLength(1);
});

// https://github.com/vant-ui/vant/issues/8889
test('should format v-model with step correctly', async () => {
  const wrapper = mount(Slider, {
    props: {
      min: 29,
      max: 39,
      step: 2,
      modelValue: 30.5,
    },
  });

  await later();
  expect(wrapper.emitted('update:modelValue')![0]).toEqual([31]);
});

test('should render left-button、right-button slot correctly', async () => {
  const wrapper = mount(Slider, {
    props: {
      range: true,
      modelValue: [30, 80],
    },
    slots: {
      'left-button': ({ value }) => `left-${value}`,
      'right-button': ({ value }) => `right-${value}`,
    },
  });

  await later();
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render reversed slider correctly', () => {
  const wrapper = mount(Slider, {
    props: {
      reverse: true,
      modelValue: 25,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render reversed vertical slider correctly', () => {
  const wrapper = mount(Slider, {
    props: {
      reverse: true,
      vertical: true,
      modelValue: 25,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render reversed range slider correctly', () => {
  const wrapper = mount(Slider, {
    props: {
      range: true,
      reverse: true,
      modelValue: [25, 40],
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should update modelValue correctly after clicking the reversed slider', () => {
  const wrapper = mount(Slider, {
    props: {
      reverse: true,
      modelValue: 50,
    },
  });

  trigger(wrapper, 'click', 100, 0);
  expect(wrapper.emitted('update:modelValue')!.pop()).toEqual([0]);
});

test('should update modelValue correctly after clicking the reversed vertical slider', () => {
  const wrapper = mount(Slider, {
    props: {
      reverse: true,
      vertical: true,
      modelValue: 50,
    },
  });

  trigger(wrapper, 'click', 0, 100);
  expect(wrapper.emitted('update:modelValue')!.pop()).toEqual([0]);
});
