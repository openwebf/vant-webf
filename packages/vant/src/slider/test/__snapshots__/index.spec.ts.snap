// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render left-button„ÄÅright-button slot correctly 1`] = `
<div class="van-slider">
  <div class="van-slider__bar"
       style="width: 50%; left: 30%;"
  >
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--left"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="30"
         aria-valuemax="100"
         aria-orientation="horizontal"
    >
      left-30
    </div>
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--right"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="80"
         aria-valuemax="100"
         aria-orientation="horizontal"
    >
      right-80
    </div>
  </div>
</div>
`;

exports[`should render readonly Slider correctly 1`] = `
<div class="van-slider">
  <div class="van-slider__bar"
       style="width: 50%; left: 0%;"
  >
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--right"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="50"
         aria-valuemax="100"
         aria-readonly="true"
         aria-orientation="horizontal"
    >
      <div class="van-slider__button">
      </div>
    </div>
  </div>
</div>
`;

exports[`should render reversed range slider correctly 1`] = `
<div class="van-slider">
  <div class="van-slider__bar"
       style="width: 15%; right: 25%;"
  >
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--left"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="25"
         aria-valuemax="100"
         aria-orientation="horizontal"
    >
      <div class="van-slider__button">
      </div>
    </div>
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--right"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="40"
         aria-valuemax="100"
         aria-orientation="horizontal"
    >
      <div class="van-slider__button">
      </div>
    </div>
  </div>
</div>
`;

exports[`should render reversed slider correctly 1`] = `
<div class="van-slider">
  <div class="van-slider__bar"
       style="width: 25%; right: 0%;"
  >
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--left"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="25"
         aria-valuemax="100"
         aria-orientation="horizontal"
    >
      <div class="van-slider__button">
      </div>
    </div>
  </div>
</div>
`;

exports[`should render reversed vertical slider correctly 1`] = `
<div class="van-slider van-slider--vertical">
  <div class="van-slider__bar"
       style="height: 25%; bottom: 0%;"
  >
    <div role="slider"
         class="van-slider__button-wrapper van-slider__button-wrapper--left"
         tabindex="0"
         aria-valuemin="0"
         aria-valuenow="25"
         aria-valuemax="100"
         aria-orientation="vertical"
    >
      <div class="van-slider__button">
      </div>
    </div>
  </div>
</div>
`;
